Utils/hydro_enforce_inout_solvability.cpp

```diff
@@ -0,0 +1,307 @@
+/** \addtogroup Utilities
+ * @{
+ */
+
+#include <hydro_utils.H>
+
+using namespace amrex;
+
+namespace HydroUtils {
+
+namespace {
+
+void set_inout_masks(
+    const int lev,
+    const Vector<Array<MultiFab*, AMREX_SPACEDIM>>& vels_vec,
+    Array<iMultiFab, AMREX_SPACEDIM>& inflow_masks,
+    Array<iMultiFab, AMREX_SPACEDIM>& outflow_masks,
+    const BCRec* bc_type,
+    const Box& domain,
+    const bool corners)
+{
+    for (OrientationIter oit; oit != nullptr; ++oit) {
+        const auto ori = oit();
+        const auto side = ori.faceDir();
+        const int dir = ori.coordDir();
+        const auto islow = ori.isLow();
+        const auto ishigh = ori.isHigh();
+
+        // Multifab for normal velocity
+        auto& vel_mf = vels_vec[lev][dir];
+
+        // mask iMFs for the respective velocity direction
+        auto& inflow_mask = inflow_masks[dir];
+        auto& outflow_mask = outflow_masks[dir];
+
+        IndexType::CellIndex dir_index_type = (vel_mf->ixType()).ixType(dir);
+        // domain extent indices for the velocities
+        int dlo;
+        if (dir_index_type == IndexType::CellIndex::CELL) {
+            // lower boundary is at -1 for cell-centered velocity
+            dlo = domain.smallEnd(dir) - 1;
+        } else {
+            // lower boundary is at  0 for face-centered velocity
+            dlo = domain.smallEnd(dir);
+        }
+        int dhi = domain.bigEnd(dir) + 1;
+
+        // get BCs for the normal velocity and set the boundary index
+        // based on low or high side
+        const BCRec ibcrec = bc_type[dir];
+        int bc, bndry;
+        if (side == Orientation::low) {
+            bc = ibcrec.lo(dir);
+            bndry = dlo;
+        } else {
+            bc = ibcrec.hi(dir);
+            bndry = dhi;
+        }
+
+        // limit influx/outflux calculations to the in-out boundaries only
+        if (bc == BCType::direction_dependent) {
```

Comment:
When I first saw "direction_dependent", it wasn't clear to me what this meant. The rest of the BCTypes are mathematical BCs. Why is "direction_dependent" the mathematical BC here instead of "mixed" or "Robin"?

Reply:
We could remove it from amrex::BCType and use `user_*`.

Reply:
`mixed` might be okay too. But What `mixed` means is likely user dependent, whereas other types (e.g., reflect_odd, foextrap, etc.) are clear on what they mean and AMReX can fill them for the user.

Reply:
We added this math BC type in AMReX (https://github.com/AMReX-Codes/amrex/pull/3965) where the cells are to be treated differently based on in/out flow. Perhaps @asalmgren could comment on the naming.

Reply:
To amrex, that's a user type that it does not know how to handle.

Reply:
On the user side, the user could define `auto my_preferred_name = amrex::BCType::user_1;`.

Reply:
I think the reason for adding a special math type rather than just using user_1 was so that it can be used consistently across dependent codes, say, amr-wind and amrex-hydro, and so it does not conflict with other custom types that may be added in the future.

Reply:
amr-wind uses amrex-hydro. So it could use the types defined in amrex-hydro. (If it does not use amrex-hydro, then it probably does not matter.)

Reply:
If this is meant to be used with both the Nodal and MAC projections, then since the BC types for those linops differ and I don't think that can be used for this purpose.

If this could be used more generally, in additional application codes, then maybe instead of having the BCRec it could be something more like a flag that indicates whether to include the face in the sum or not?